sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala

```diff
@@ -134,10 +137,16 @@ final class Decimal extends Ordered[Decimal] with Serializable {
       // result. For example, the precision of 0.01 equals to 1 based on the definition, but
       // the scale is 2. The expected precision should be 3.
       this._precision = decimal.scale + 1
+      this._scale = decimal.scale
+    } else if (decimal.scale < 0 && !SQLConf.get.allowNegativeScaleOfDecimalEnabled) {
+      this._precision = decimal.precision - decimal.scale
```

Comment:
No chance that this `_precision` goes over the max precision?

Reply:
It's possible. But we have `DecimalType` aside to do the check and fail if we goes over the max precision.

Reply:
Yea, but I personally think it'd be better to check the max precision in the setter of a `Decimal` side, too, for safeguards. It seems the other setter has the check: https://github.com/apache/spark/blob/1846b0261b84ce1bca079bc59fb4518bff910c18/sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala#L117.

Reply:
if we add check for this, do we also have to add check for https://github.com/apache/spark/blob/1846b0261b84ce1bca079bc59fb4518bff910c18/sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala#L138 ?

Reply:
+1

Reply:
yea, I think so.